docs/spec/constructors.rst

```diff
@@ -0,0 +1,487 @@
+
+For most classes, the return type for the ``__new__`` method is typically
+``Self``, but other types are also allowed. For example, the ``__new__``
+method may return an instance of a subclass or an instance of some completely
+unrelated class.
+
+If the evaluated return type of ``__new__`` is not an instance of the class
+being constructed (or a subclass thereof) or is a union that includes such
+a class, a type checker should assume that the ``__init__`` method will not be
+called. This is consistent with the runtime behavior of the ``type.__call__``
+method.
+
+  ::
+
+    class MyClass:
+        def __new__(cls) -> int:
+            return 0
+
+        # The __init__ method will not be called in this case, so
+        # it should not be evaluated.
+        def __init__(self, x: int):
+            pass
+
+    assert_type(MyClass(), int)
+
+For purposes of this test, an explicit return type of ``Any`` (or a
+union containing ``Any``) should be treated as a type that is not an instance
+of the class being constructed.
+
+  ::
+
+    class MyClass:
+        def __new__(cls) -> Any:
+            return 0
+
+        # The __init__ method will not be called in this case, so
+        # it should not be evaluated.
+        def __init__(self, x: int):
+            pass
+
+    assert_type(MyClass(), Any)
+
+If the return type of ``__new__`` is not annotated, a type checker may assume
+that the return type is ``Self`` and proceed with the assumption that the
+``__init__`` method will be called.
+
+If the class is generic, it is possible for a ``__new__`` method to override
+the specialized class type and return a class instance that is specialized
+with different type arguments.
+
+  ::
+
+    class MyClass[T]:
+        def __new__(cls, *args, **kwargs) -> "MyClass[list[T]]":
+            ...
+
+    assert_type(MyClass[int](), MyClass[list[int]])
+
+If the ``cls`` parameter within the ``__new__`` method is not annotated, type
+checkers should infer a type of ``type[Self]``. Regardless of whether the
+type of the ``cls`` parameter is explicit or inferred, the type checker should
+bind the class being constructed to this parameter and report any type errors
+that arise during binding.
+
+  ::
+
+    class MyClass[T]:
+        def __new__(cls: "type[MyClass[int]]") -> "MyClass[int]": ...
+
+    MyClass()  # OK
+    MyClass[int]()  # OK
+    MyClass[str]()  # Type Error
+
+
+``__init__`` Method
+===================
+
+After evaluating the ``__new__`` method, a type checker should evaluate the
+``__init__`` method (if applicable) using the supplied arguments. If the class
+is generic and explicitly specialized (or specialized via the ``__new__`` method
+return type), the type checker should partially specialize the ``__init__``
+method using the supplied type arguments. If the class is not explicitly
+specialized, class-scoped type variables should be solved using the supplied
+arguments passed to the constructor call.
```

Comment:
If the method includes both class-scoped and method-scoped type variables, the class-scoped type variables can be provided through explicit specialization. The method-scoped type variables cannot be explicitly specialized.

It's also possible that the `__new__` method returns a type that supplies specialized values for a subset of class-scoped type variables. The `__init__` method can then be used to infer the values of the remaining type variables.

```python
class MyDict[K, V](dict[K, V]):
    def __new__(cls, *args, **kwargs) -> MyDict[str, V]: ...
    def __init__(self, value: V) -> None: ...

reveal_type(MyDict[int, int](3)) # Should reveal "MyDict[str, int]"
```

You might ask "why would anyone do this"? There appears to be use cases for it because I've seen things like this many times, even in mainstream libraries like pandas.